### PR TITLE
feat: `plugin::Builder` closes #2959

### LINKED ISSUE #2959: [Proposal]: Plugin Builder
> This proposal is not finished yet! Please participate by reviewing the proposed API.

**Is your feature request related to a problem? Please describe.**

There are several observations I made about the current `Plugin` trait that I would like to address:
1. Defining a custom struct for a plugin has little realworld use as plugin state is better stored using the `app.manage` and `app.state` methods. (This state is also accessible in commands while plugin struct fields are not)
2. the `Plugin` method names are not super self explanatory and naming is somewhat inconsistent (`created` is an event callback but is not named `on_created` while `on_event` and `on_page_load` are)
3. The practice of having an `invoke_handler` field and a `extend_api` to register commands is not obvious
4. In "real-world" usage (the official plugins) the `config` property taken from the `tauri.conf.json` file had little use and makes plugins more confusing rather than easier to use. It has shown that exporting a builder and having configuration "in-code" is better.

In general I think the current Plugin api leads to unnecessarily verbose and complex plugins, so I would like to propose an improved builder based API:

**Describe the solution you'd like**
Introduce a `PluginBuilder` struct to streamline the plugin creation:
```rs
/// Mock command
#[command]
async fn execute<R: Runtime>(_app: AppHandle<R>) -> Result<String> {
  Ok("success".to_string())
}

pub fn init() -> Plugin { // not the Plugin trait
   PluginBuilder::new("example")
      .commands(vec![execute])
      .setup(|app| {
         app.manage(MyState::default());
         Ok(())
      })
}
```

Instead of the current api:
```rs
/// Mock command
#[command]
async fn execute<R: Runtime>(_app: AppHandle<R>) -> Result<String> {
  Ok("success".to_string())
}

pub struct ExamplePlugin<R: Runtime> {
  invoke_handler: Box<dyn Fn(Invoke<R>) + Send + Sync>,
}

impl<R: Runtime> Default for ExamplePlugin<R> {
  fn default() -> Self {
    Self {
      invoke_handler: Box::new(tauri::generate_handler![execute]),
    }
  }
}

impl<R: Runtime> Plugin<R> for ExamplePlugin<R> {
  fn name(&self) -> &'static str {
    "example"
  }

  fn initialize(&mut self, app: &AppHandle<R>, _config: JsonValue) -> tauri::plugin::Result<()> {
    app.manage(MyState::default());
    Ok(())
  }

  fn extend_api(&mut self, message: Invoke<R>) {
    (self.invoke_handler)(message)
  }
}
```

The proposed builder methods closely resemble the traits methods, with a few differences:
- rename methods to better communicate their purpose 
   `js` instead of `initiailization_script`, `commands` instead of `extend_api` and `setup` instead of `initialize`.
- name event callbacks consistently
  with the prefix `on_`

```rs
trait PluginBuilder {
   // creates the builder using the plugin `name`
   fn new(name: &'static str) -> Self;
   // Renamed `initialization_script` fn to better communicate it takes a JavaScript string
   fn js(&mut self, js_code: String) -> &mut self;
   // Register an array of tauri commands. This hides all the ugly Boxing and macro usage
   fn commands(&mut self, commands: Vec<?>) -> &mut self;
   // Renamed `initialize` fn to highlight the similarity to the `AppBuilder`s setup fn
   fn setup(setup: Fn(&AppHandle<R>) -> Result<(), Box<dyn Error + Send>> + Send + 'static);
   // set the `on_event ` method
   fn on_event(cb: Fn(&AppHandle<R>, &Event)) -> &mut self;
   // set the `on_page_load` method
   fn on_page_load(cb: Fn(Window<R>, payload: PageLoadPayload)) -> &mut self;
   // set the `created` method
   fn on_webview_ready(cb: Fn(Window<R>)) -> &mut self;
   // builds the plugin
   fn build(self) -> Plugin; // not the Plugin trait
}
```

The `PluginBuilder` struct could just be wrapper returning `Plugin` trait implementations, so this proposal would have a clear implementation path.

### Notes

1. Having a builder struct allows us to improve and extend the plugin API in the future with less hassle
2. I propose plugins exporting a creation function called `init` (as shown in the example above). This function is free to receive any arguments, so plugin authors may use this to provide configuration options for end users.
3. I'm considering to remove the `js`/`initialization_script` methods all together. Plugins should expose corresponding js packages and not rely on extending the global object. The overwhelming majority of tauri projects also uses bundlers which make providing api packages is a no-brainer. For the few projects that don't, plugin can provide prebuilt *browser bundles*. See [tauri-plugin-store](https://github.com/tauri-apps/tauri-plugin-store/blob/dev/webview-src/rollup.config.js) for an example of this.

**Describe alternatives you've considered**
Leave the trait as is.

**Additional context**
The builder pattern is inspired by [`deno_core`s `ExtensionBuilder`](https://github.com/denoland/deno/blob/main/core/extensions.rs#L61-L108).
The `init`function is inspired by [denos extensions](https://github.com/denoland/deno/tree/main/ext) and the convention among [rollup](https://rollupjs.org) plugins to export a [single creation function](https://rollupjs.org/guide/en/#a-simple-example).

### PR DESCRIPTION
**Pull request by JonasKruckenberg**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding Issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes. Issue #___
- [x] No

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)
see #2959

**Other information:**

### CODE REVIEW

**Comment by JonasKruckenberg** [comment]
Alright this PR is ready to be reviewed and merged!

File: core/tauri/src/plugin.rs [hunk 1]
```diff
@@ -52,6 +53,145 @@ pub trait Plugin<R: Runtime>: Send {
...
+    self.invoke_handler = Box::new(invoke_handler);
+    self
+  }
+
+  /// The JS script to evaluate on webview initialization.
+  /// The script is wrapped into its own context with `(function () { /* your script here */ })();`,
+  /// so global variables must be assigned to `window` instead of implicity declared.
+  ///
+  /// It's guaranteed that this script is executed before the page is loaded.
+  pub fn initialization_script(mut self, initialization_script: String) -> Self {
```
Comment (JonasKruckenberg) [inline core/tauri/src/plugin.rs]: What do you think about `js_init_script` @probablykasper proposed?

Comment (amrbashir) [reply core/tauri/src/plugin.rs]: I like it

File: core/tauri/src/plugin.rs [hunk 2]
```diff
@@ -52,6 +53,145 @@ pub trait Plugin<R: Runtime>: Send {
   fn extend_api(&mut self, invoke: Invoke<R>) {}
 }
 
+type SetupHook<R> = dyn Fn(&AppHandle<R>) -> Result<()> + Send + Sync;
+type OnWebviewReady<R> = dyn Fn(Window<R>) + Send + Sync;
+type OnEvent<R> = dyn Fn(&AppHandle<R>, &Event) + Send + Sync;
+
+/// Builds a tauri plugin
```
Comment (amrbashir) [inline core/tauri/src/plugin.rs]:
```suggestion
/// Builds a [generic tauri plugin](GenericPlugin)
```

File: core/tauri/src/plugin.rs [hunk 3]
```diff
@@ -52,6 +53,145 @@ pub trait Plugin<R: Runtime>: Send {
...
+  /// Callback invoked when the event loop receives a new event.
+  pub fn on_event<F>(mut self, on_event: F) -> Self
+  where
+    F: Fn(&AppHandle<R>, &Event) + Send + Sync + 'static,
+  {
+    self.on_event = Box::new(on_event);
+    self
+  }
+
+  /// Builds the plugin.
```
Comment (amrbashir) [inline core/tauri/src/plugin.rs]:
```suggestion
  /// Builds the [generic plugin](GenericPlugin).
```

File: docs/usage/guides/plugin-builder.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
+---
+title: Write Tauri Plugins
+---
+
+import Alert from '@theme/Alert'
+
+<Alert title="Note" icon="info-alt">
+This guide uses the `plugin::Builder` introduced in `v1`, if you're looking for the old version using trait implementations [see here](./plugin.md)
```
Comment (amrbashir) [inline docs/usage/guides/plugin-builder.md]: Can't we combine both? maybe improve the old guide and add it under **Advanced** section or something.

File: docs/usage/guides/plugin-builder.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+</Alert>
+
+Plugins allow you to hook into the Tauri application lifecycle and introduce new commands.
+
+## Getting started
+
+Plugins are reusable extensions to the Tauri API that solve common problems. They are also a very convenient way to structure your own code base!
+
+If you intend to share you plugin with others, we provide a ready-made template! With the tauri-cli installed just run:
```
Comment (amrbashir) [inline docs/usage/guides/plugin-builder.md]:
```suggestion
If you intend to share your plugin with others, we provide a ready-made template! With the tauri-cli installed just run:
Write
```

File: docs/usage/guides/plugin-builder.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+```sh
+tauri init plugin --name awesome
```
Comment (amrbashir) [inline docs/usage/guides/plugin-builder.md]:
```suggestion
tauri plugin init --name awesome
```

File: docs/usage/guides/plugin-builder.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+By default consumers of your plugin can call provided commands like this:
+
+```ts
+import { invoke } from '@tauri-apps/api'
+
+invoke('plugin:awesome|do_something')
+```
+where `awesome` will be replaced by your plugin name.
+
+This isn't very convenient however, so it's common for plugins to prive a so called *API package*, a JavaScript package that provides convenient access to your commands.
```
Comment (amrbashir) [inline docs/usage/guides/plugin-builder.md]:
```suggestion
This isn't very convenient however, so it's common for plugins to provide a so called *API package*, a JavaScript package that provides convenient access to your commands.
```

File: docs/usage/guides/plugin-builder.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+> An example of this is the [tauri-plugin-store](https://github.com/tauri-apps/tauri-plugin-store), that provides a convenient class structure to accessing a store.
+
+You can scaffold a plugin with attached API package like this:
+
+```sh
+tauri init plugin --name awesome --api
```
Comment (amrbashir) [inline docs/usage/guides/plugin-builder.md]:
```suggestion
tauri plugin init --name awesome --api
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: docs/usage/guides/plugin-builder.md (hunk 4 above)
Comment (JonasKruckenberg) [reply docs/usage/guides/plugin-builder.md]: yeah you're right.

Comment (JonasKruckenberg) [reply docs/usage/guides/plugin-builder.md]: fixed!

File: core/tauri/src/plugin.rs (hunk 1 above)
Comment (JonasKruckenberg) [reply core/tauri/src/plugin.rs]: Changed to `js_init_script`!

**Review by amrbashir — APPROVED** [review]

**Comment by JonasKruckenberg** [comment]
@lucasfernog Idk why the checks have failed, but this PR is good for merging regardless

**Pull request merged**